src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -168,13 +169,89 @@ private void OnEventSourceChanged(object? _1, TaggerEventArgs _2)
             => EnqueueWork(highPriority: false);
 
         private void EnqueueWork(bool highPriority)
-            => _eventChangeQueue.AddWork(highPriority, _dataSource.CancelOnNewWork);
+            => EnqueueWork(highPriority, _dataSource.SupportsFrozenPartialSemantics);
 
-        private ValueTask<VoidResult> ProcessEventChangeAsync(ImmutableSegmentedList<bool> changes, CancellationToken cancellationToken)
+        private void EnqueueWork(bool highPriority, bool frozenPartialSemantics)
         {
+            // Cancellation token if this expensive work that we want to be cancellable when cheap work comes in.
+            CancellationToken? nonFrozenComputationToken = null;
+
+            if (_dataSource.SupportsFrozenPartialSemantics)
+            {
+                // We do support frozen partial work.  Cancel any expensive work in flight as new work has come in.
+                var nextToken = _nonFrozenComputationCancellationSeries.CreateNext();
+
+                // If this is new work *is* the expensive work, then use this token to allow cancellation of it when
+                // more new work comes in.
+                if (!frozenPartialSemantics)
+                    nonFrozenComputationToken = nextToken;
+            }
+
+            _eventChangeQueue.AddWork(
+                new TagSourceQueueItem(highPriority, frozenPartialSemantics, nonFrozenComputationToken),
+                _dataSource.CancelOnNewWork);
+        }
+
+        private async ValueTask ProcessEventChangeAsync(
+            ImmutableSegmentedList<TagSourceQueueItem> changes, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfTrue(changes.IsEmpty);
+
             // If any of the requests was high priority, then compute at that speed.
-            var highPriority = changes.Contains(true);
-            return new ValueTask<VoidResult>(RecomputeTagsAsync(highPriority, cancellationToken));
+            var highPriority = changes.Contains(x => x.HighPriority);
```

Comment:
`Contains` is not optimized for lambda callbacks, but `Any` is:

```suggestion
            var highPriority = changes.Any(x => x.HighPriority);
```

Reply:
well that's silly :D